### ISSUE TITLE
Derivatives/residuals

### ISSUE THREAD

**Issue by joshday** [JuliaML/Losses.jl#26]
I was adding a QuantileLoss type when I noticed this.  In statistics, `residual = y - yhat`.  Our distance losses are implemented as if `residual = yhat - y`, so derivatives are missing a minus sign.

```
# L(y, yhat) = abs(y - yhat)
value(loss::L1DistLoss, residual) = abs(residual)
deriv(loss::L1DistLoss, residual) = sign(residual)  # missing a minus sign
```

There's several places in the code with `h(x) - y` (plot recipes, docs).  Are there reasons for this, or can we switch to statistical notation?

**Comment by Evizero** [comment]
I did this on purpose because otherwise the derivative would be inverse to what one expects it to be (so suddenly one has to add the gradient instead of subtracting it during learning). Something that the plots as well as testing against automatic differentiation showed. This way all works out nicely and is also in line with typical notation used in ML classes (such as Prof. Ng's). Although I do admit that it does go against the notation from the Loss literature.

In other words, I took the easy path.

**Comment by Evizero** [comment]
I will also admit that I did not spend a substantial amount of time investigating the root cause, as I am sure there must be a reason for the Loss literature (and apparently statistics) to go that path. As soon as I discovered that changing the order solved all my problems I accepted that.

Maybe you can explain to me why this order is preferred ?

**Comment by joshday** [comment]
In statistics, `residual = actual - predicted` is a definition, not just convention.  It's a part of the history of linear regression.  The choice is arbitrary, but I think it could keep statisticians from buying in to JuliaML.  I remember looking at EmpiricalRisks.jl and thinking everything is backwards (and part of the reason I used my own implementations in my packages).

**Comment by Evizero** [comment]
As long as the derivatives work out as they should I don't mind either way. Maybe I had some implementation errors or something

**Comment by tbreloff** [comment]
I don't remember any strong arguments for either form, but consistency is
key. Just make sure we do it the same everywhere and I'm happy.

**Comment by ahwillia** [comment]
Optimization literature seems to use the opposite convention. For example, solving linear least squares is always written as:

```
minimize_x  ||Ax - b||
```

I think the reason is what @Evizero was saying. You get a sign flip by chain rule when you differentiate  `||b - Ax||` w.r.t. `x`. 

I don't think one convention is better than the other (though maybe we shouldn't use the word `residual` as this is a misnomer). As long as they have similar performance, I'm happy with either.

**Comment by joshday** [comment]
There's no sign change in those derivatives with respect to `x`: `Ax - b == -(b - Ax)`.  I believe our issue is that we want the convenience of treating DistanceLoss as a function of `residual` alone.  We're not thinking of `f(x)`, we're thinking of `f(Ax - b)`.  If instead we do `f(b - Ax)`, we would need to add a minus sign for `deriv`.  

I say we just find a better term for residual, although I don't have any suggestions that immediately come to mind.

**Comment by tbreloff** [comment]
I'm not convinced we need special residual versions at all.  I agree with @joshday that it hides what we're really doing.

**Comment by Evizero** [comment]
> I say we just find a better term for residual, although I don't have any suggestions that immediately come to mind.

This seems like my preferred option

**Comment by ahwillia** [comment]
Adding the minus sign in front of deriv was what I meant.

+1 for just substituting a new word for resid. Or does it even need a word? `r = A*x - b`

**Comment by ahwillia** [comment]
Adding the minus sign in deriv was what I meant.

+1 for just substituting a new word for resid. Or does it even need a word? `r = A*x - b`

**Comment by joshday** [comment]
`r` works for me.

> I'm not convinced we need special residual versions at all

I can imagine cases where you might already have a vector of residuals and don't want to recalculate them to get the loss.  @Evizero Was this your use case?  

I think I only need the `value(loss, y, yhat)` api, but I can see the argument for residual versions.

**Comment by Evizero** [comment]
aside from being a nice interface for educational purposes in exercises, code beauty and simplicity is my only use case in this. It does make more sense to me as a non statistician to not have to augment the derivative with a minus just to make things work out as expected.

I very much understand that the apparent name misuse is annoying/confusing and I will fix that

**Comment by Evizero** [comment]
I am fixing this right now. I shall call it `difference`, which is literally what it is and is more in line with the recent spirit of using descriptive names instead of short acronyms

**Comment by joshday** [comment]
> code beauty and simplicity is my only use case in this

If that's the case, can we drop the DistanceLoss/MarginLoss abstract types?  Separating them makes it harder to write algorithms that work with both.  It also adds this residual/difference confusion.

**Comment by Evizero** [comment]
No, I need the types for distinction between classification and regression. how does it make it more difficult? they share the same interface plus some more specialized signatures. or at least that is what should be the case, maybe I missed something. Can you give me an example for your problem?

**Comment by Evizero** [comment]
The code beauty comment was about the types having this simplification for their individual implementation. The existence of the types on the other hand was a major reason I decided not to use the `EmpiricalRisks` package, but implement my own instead

**Comment by joshday** [comment]
> No, I need the types for distinction between classification and regression

I still don't understand how you're using them.  

So it doesn't make it harder, but we're adding extra code to give them the same api.  Why not just start with that api?

**Comment by Evizero** [comment]
well, in the case of SVM at least, the only thing that is always able to tell you what kind of problem you are solving are the Loss functions. The data itself can be ambiguous (-1 and 1 could also be regression output, and one might not want to check all the labels to make a decision), the solver can be ambiguous (gradient descent, pegasos). So if one want to have a type stable high-level interface that doesn't poison the type inference then the solution is to either force a clear separation of regression and classification or make use of the information in the loss.
To me it seems like a redundant information that if I take a hinge loss, I still have to say it is a classification problem.

Furthermore I want to be able to dispatch on the loss. Some algorithms only work with a certain kind of loss (let's say margin-based) and need further information for that loss. Information that is undefined for distance-based losses, for example if the loss is clipable.
Other algorithm work for both kind of losses but have slight differences, so you would want to have distinct implementations for margin and distance based. 

Last but not least one would want to have sensible fallback methods and those are only really possible at compile time.

another thing I would like is that a high-level interface is able to choose the return structure based on its parameters (and is still type stable). So if it knows I have a classification problem it gives me a convenient structure back that is specialized for classification. If I don't offer that possibility low-level then a high-level interface has to be type-unstable by definition.

Another usecase that I like is for educational purposes. If one wants his/her students to explore loss functions then this package in the current state has a lot to offer in my opinion. you can query Losses in a way that one typically sees them plotted, and get a feeling of the connection of the loss plots (which are plots of one variable) and the usual loss interface, which have two variables.

The implementation for each loss is simplified and cleaner as well.

You are not the first to raise the proposal, @tbreloff suggested it a couple of times. I still fail to see why it is so desirable to remove something that to me seems so useful, and maybe more importantly harmless to anyone who wants to opt out.

**Comment by Evizero** [comment]
But look, in the end I want this to be a community package, so I am willing to put it to a vote and respect it's outcome. I would then try to make it work for me with the voted outcome. But please don't take this away from me without a reason other than _less code_, which in this case to me just translates to _less functionality_.

**Comment by joshday** [comment]
Suppose we keep the abstract types and remove the specialized difference/agreement methods in favor of a `fun(loss, actual, predicted)` api?  This is just as functional, cleaner, and easier for an outsider to understand.

**Comment by Evizero** [comment]
this api is available though? or do you mean refactor the individual losses to be implemented that way?

**Comment by joshday** [comment]
I mean refactor.

**Comment by Evizero** [comment]
because refactoring has an ugly overhead (i.e. it is not as clean). this way we allow without any further work that `actual` and `predicted` don't have to be of the same type. if we refactor we would either have to restrict to those variables to be of the same type, or add a lot of type promotion uglyness

**Comment by Evizero** [comment]
I suppose we could implement the losses to be restricted to have the same types, but implement a fallback method that does type promotion in the case that the types differ. It would not be as efficient as it is currently (which creates special implementations for the specified type combination) but it would offer the same flexibility. The argument could be made that if efficiency matters that one should use the same types to begin with, although it is always annoying when working with svms that the first thing one has to do is convert all labels from integers to floats

**Comment by Evizero** [comment]
Personally I am of the opinion that it is cleaner with the agreement/difference implementation, but I could very well live with the refactor compromise if need be. Let's get @tbreloff and @ahwillia opinion on this and I will act accordingly

**Comment by tbreloff** [comment]
I think my opinion is that as long as you can access the generic api that
it's fine to have the more specific one. Especially since it's already
built.

**Comment by ahwillia** [comment]
I think I agree with @tbreloff and don't quite see why we would have to change what we have currently. I might be convinced with more specifics, but this doesn't seem high priority to me at the moment.

**Comment by Evizero** [comment]
@joshday Is this something that is important to you / your adaption? If not I will leave this as is for now, given the general mood of the answers. If it is something that you consider a major obstacle for your opt-in into the package I am still willing to make the change as a compromise

**Comment by joshday** [comment]
Not an obstacle, mainly just bikeshedding.  As long as we have the generic api I'm good.
